File: src/test/java/com/huawei/wallet/hms/eventticket/EventTicketInstanceTest.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.wallet.hms.eventticket;

import com.huawei.wallet.hms.ServerApiService;
import com.huawei.wallet.hms.ServerApiServiceImpl;
import com.huawei.wallet.util.ConfigUtil;
import com.huawei.wallet.util.HwWalletObjectUtil;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;

import org.junit.Test;

/**
 * Event ticket instance tests.
 *
 * @since 2019-12-12
 */
public class EventTicketInstanceTest {
    private final ServerApiService serverApiService = new ServerApiServiceImpl();

    /**
     * Add an event ticket instance to HMS wallet server.
     * Run the "createEventTicketModel" test before running this test.
     * After using this API, you will use a thin JWE to bind this instance to a user. Or you can add an instance by
     * sending a JWE with complete instance information, without using this API. See JWE example tests.
     * POST http://XXX/hmspass/v1/eventticket/instance
     */
    @Test
    public void addEventTicketInstance() {
        System.out.println("addEventTicketInstance begin.");

        // Read an event ticket instance from a JSON file.
        JSONObject instance = JSONObject.parseObject(ConfigUtil.readFile("EventTicketInstance.json"));

        // Validate parameters.
        HwWalletObjectUtil.validateInstance(instance);

        // Post the new event ticket instance to HMS wallet server.
        String urlSegment = "/v1/eventticket/instance";
        JSONObject responseInstance =
            serverApiService.postToWalletServer(urlSegment, JSONObject.toJSONString(instance));
        System.out.println("Posted event ticket instance: " + JSONObject.toJSONString(responseInstance));
    }

    /**
     * Get an event ticket instance by its instance ID.
     * Run the "createEventTicketInstance" test before running this test.
     * GET http://xxx/hmspass/v1/eventticket/instance/{instanceId}
     */
    @Test
    public void getEventTicketInstance() {
        System.out.println("getEventTicketInstance begin.");

        // ID of the event ticket instance you want to get.
        String instanceId = "EventTicketPass10001";

        // Get the event ticket instance.
        String urlSegment = "/v1/eventticket/instance/";
        JSONObject responseInstance = serverApiService.getHwWalletObjectById(urlSegment, instanceId);
        System.out.println("Corresponding event ticket instance: " + JSONObject.toJSONString(responseInstance));
    }

    /**
     * Get event ticket instances belonging to a specific event ticket model.
     * Run the "createEventTicketInstance" test before running this test.
     * GET http://xxx/hmspass/v1/eventticket/instance?modelId=XXX&session=XXX&pageSize=XXX
     */
    @Test
    public void getEventTicketInstanceList() {
        System.out.println("getEventTicketInstanceList begin.");

        // Model ID of event ticket instances you want to get.
        String modelId = "EventTicketTestModel";

        // Get a list of the event ticket instances.
        String urlSegment = "/v1/eventticket/instance";

        JSONArray instances = serverApiService.getInstances(urlSegment, modelId, 5);
        System.out.println("Total instances count: " + instances.size());
        System.out.println("Instances list: " + instances.toJSONString());
    }

    /**
     * Overwrite an event ticket instance.
     * Run the "createEventTicketInstance" test before running this test.
     * PUT http://xxx/hmspass/v1/eventticket/instance/{instanceId}
     */
    @Test
    public void fullUpdateEventTicketInstance() {
        System.out.println("fullUpdateEventTicketInstance begin.");

        // Read a HwWalletObject from a JSON file. This HwWalletObject will overwrite the corresponding instance.
        JSONObject instance = JSONObject.parseObject(ConfigUtil.readFile("FullUpdateEventTicketInstance.json"));

        // Validate parameters.
        HwWalletObjectUtil.validateInstance(instance);

        // Update the event ticket instance.
        String urlSegment = "/v1/eventticket/instance/";
        JSONObject responseInstance = serverApiService.fullUpdateHwWalletObject(urlSegment,
            instance.getString("serialNumber"), JSONObject.toJSONString(instance));
        System.out.println("Updated event ticket instance: " + JSONObject.toJSONString(responseInstance));
    }

    /**
     * Update an event ticket instance.
     * Run the "createEventTicketInstance" test before running this test.
     * PATCH http://xxx/hmspass/v1/eventticket/instance/{instanceId}
     */
    @Test
    public void partialUpdateEventTicketInstance() {
        System.out.println("partialUpdateEventTicketInstance begin.");

        // ID of the event ticket instance you want to update.
        String instanceId = "EventTicketPass10001";

        // Read a HwWalletObject from a JSON file. This HwWalletObject will merge with the corresponding instance.
        String instanceStr = ConfigUtil.readFile("PartialUpdateEventTicketInstance.json");

        // Update the event ticket instance.
        String urlSegment = "/v1/eventticket/instance/";
        JSONObject responseInstance = serverApiService.partialUpdateHwWalletObject(urlSegment, instanceId, instanceStr);
        System.out.println("Updated event ticket instance: " + JSONObject.toJSONString(responseInstance));
    }

    /**
     * Add messages to an event ticket instance.
     * Run the "createEventTicketInstance" test before running this test.
     * POST http://xxx/hmspass/v1/eventticket/instance/{instanceId}/addMessage
     */
    @Test
    public void addMessageToEventTicketInstance() {
        System.out.println("addMessageToEventTicketInstance begin.");

        // ID of the event ticket instance you want to update.
        String instanceId = "EventTicketPass10001";

        // Create a list of messages you want to add to an instance. Each message contains key, value, and label.
        // The list should not contain multiple messages with the same key. You can update an existing message by adding
        // a new message with the same key. One instance contains at most 10 messages. If an instance already have 10
        // messages and you keep adding new messages, the oldest messages will be removed. You should not add more than
        // 10 messages at a time.

        // Read messages from a JSON file.
        String messagesStr = ConfigUtil.readFile("Messages.json");

        // Add messages to the event ticket instance.
        String urlSegment = "/v1/eventticket/instance/";
        JSONObject responseInstance = serverApiService.addMessageToHwWalletObject(urlSegment, instanceId, messagesStr);
        System.out.println("Updated event ticket instance: " + JSONObject.toJSONString(responseInstance));
    }
}

File: src/demo/eventticket/eventticketinstance_test.go
package eventticket

import (
	"../../../src/api"
	"../../model"
	"fmt"
	"io/ioutil"
	"path/filepath"
	"testing"
)


// Add an event ticket instance to HMS wallet server.
// Run the "createEventTicketModel" test before running this test.
// After using this API, you will use a thin JWE to bind this instance to a user. Or you can add an instance by
// sending a JWE with complete instance information, without using this API. See JWE example tests.
// POST http://XXX/hmspass/v1/eventticket/instance
func TestAddEventTicketInstance(t *testing.T) {
	t.Log("addEventTicketInstance begin.")

	// Read an event ticket instance from a JSON file.
	dataPath, _ := filepath.Abs("../data/EventTicketInstance.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	// Post the new event ticket instance to HMS wallet server.
	urlSegment := "eventticket/instance"
	resp, err := api.PostToWalletServer(urlSegment, dataByte)
	if err != nil {
		t.Logf("Add event ticket instance failed. %v", err)
	} else {
		t.Logf("Add event ticket instance success, eventticket/instance response: %v", resp)
	}
}

// Get an event ticket instance by its instance ID.
// Run the "createEventTicketInstance" test before running this test.
// GET http://xxx/hmspass/v1/eventticket/instance/{instanceId}
func TestGetEventTicketInstance(t *testing.T) {
	t.Log("getEventTicketInstance begin.")

	// ID of the event ticket instance you want to get.
	instanceId := "EventTicketPass10001"

	// Get the event ticket instance.
	urlSegment := fmt.Sprintf("eventticket/instance/%s", instanceId)
	resp, err := api.GetHwWalletObjectById(urlSegment)
	if err != nil {
		t.Logf("Get event ticket instance failed. %v", err)
	} else {
		t.Logf("Get event ticket instance success, instanceId: %s, response: %v", instanceId, resp)
	}
}

// Get event ticket instances belonging to a specific event ticket model.
// Run the "createEventTicketInstance" test before running this test.
// GET http://xxx/hmspass/v1/eventticket/instance?modelId=XXX&session=XXX&pageSize=XXX
func TestGetEventTicketInstanceList(t *testing.T) {
	t.Log("getEventTicketInstanceList begin.")

	// Model ID of event ticket instances you want to get.
	modelId := "EventTicketTestModel"

	// Get the event ticket instance.
	urlSegment := "eventticket/instance"
	api.GetInstances(urlSegment, modelId, 5)
}

// Update an event ticket instance.
// Run the "createEventTicketInstance" test before running this test.
// PATCH http://xxx/hmspass/v1/eventticket/instance/{instanceId}
func TestFullUpdateEventTicketInstance(t *testing.T) {
	t.Log("fullUpdateEventTicketInstance begin.")

	// Read a HwWalletObject from a JSON file. This HwWalletObject will overwrite the corresponding instance.
	dataPath, _ := filepath.Abs("../data/FullUpdateEventTicketInstance.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	var instance model.HwWalletObject
	api.JSONByteToStruct(dataByte, &instance)

	urlSegment := fmt.Sprintf("eventticket/instance/%s", instance.SerialNumber)

	resp, err := api.FullUpdateHwWalletObject(urlSegment, dataByte)
	if err != nil {
		t.Logf("Update event ticket instance failed. %v", err)
	} else {
		t.Logf("Update event ticket instance success, response: %v", resp)
	}
}

// Update an event ticket instance.
// Run the "createEventTicketInstance" test before running this test.
// PATCH http://xxx/hmspass/v1/eventticket/instance/{instanceId}
func TestPartialUpdateEventTicketInstance(t *testing.T) {
	t.Log("partialUpdateEventTicketInstance begin.")

	// ID of the event ticket instance you want to update.
	instanceId := "EventTicketPass10001"

	// Read a HwWalletObject from a JSON file. This HwWalletObject will merge with the corresponding instance.
	dataPath, _ := filepath.Abs("../data/PartialUpdateEventTicketInstance.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	urlSegment := fmt.Sprintf("eventticket/instance/%s", instanceId)
	resp, err := api.PartialUpdateHwWalletObject(urlSegment, dataByte)
	if err != nil {
		t.Logf("Update event ticket instance failed. %v", err)
	} else {
		t.Logf("Update event ticket instance success, response: %v", resp)
	}
}

// Add messages to an event ticket instance.
// Run the "createEventTicketInstance" test before running this test.
// POST http://xxx/hmspass/v1/eventticket/instance/{instanceId}/addMessage
func TestAddMessageToEventTicketInstance(t *testing.T) {
	t.Log("addMessageToEventTicketInstance begin.")

	// ID of the event ticket instance you want to update.
	instanceId := "EventTicketPass10001"

	// Create a list of messages you want to add to an instance. Each message contains key, value, and label.
	// The list should not contain multiple messages with the same key. You can update an existing message by adding
	// a new message with the same key. One instance contains at most 10 messages. If an instance already have 10
	// messages and you keep adding new messages, the oldest messages will be removed. You should not add more than
	// 10 messages at a time.

	// Read messages from a JSON file.
	dataPath, _ := filepath.Abs("../data/Messages.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	// Add messages to the event ticket instance.
	urlSegment := fmt.Sprintf("eventticket/instance/%s/addMessage", instanceId)
	resp, err := api.AddMessageToHwWalletObject(urlSegment, dataByte)
	if err != nil {
		t.Logf("Add message to event ticket instance failed. %v", err)
	} else {
		t.Logf("Add message to event ticket instance success, response: %v", resp)
	}
}
